library/core/src/iter/adapters/flatten.rs

```diff
@@ -136,26 +136,12 @@ where
 }
 
 #[unstable(feature = "trusted_len", issue = "37572")]
-unsafe impl<T, I, F, const N: usize> TrustedLen for FlatMap<I, [T; N], F>
+unsafe impl<I, U, F> TrustedLen for FlatMap<I, U, F>
 where
-    I: TrustedLen,
-    F: FnMut(I::Item) -> [T; N],
-{
-}
-
-#[unstable(feature = "trusted_len", issue = "37572")]
-unsafe impl<'a, T, I, F, const N: usize> TrustedLen for FlatMap<I, &'a [T; N], F>
-where
-    I: TrustedLen,
-    F: FnMut(I::Item) -> &'a [T; N],
-{
-}
-
-#[unstable(feature = "trusted_len", issue = "37572")]
-unsafe impl<'a, T, I, F, const N: usize> TrustedLen for FlatMap<I, &'a mut [T; N], F>
-where
-    I: TrustedLen,
-    F: FnMut(I::Item) -> &'a mut [T; N],
+    I: Iterator,
+    U: IntoIterator,
+    F: FnMut(I::Item) -> U,
+    FlattenCompat<Map<I, F>, <U as IntoIterator>::IntoIter>: TrustedLen,
```

Comment:
I'm not sure this is correct. How does it ensure that the map function returns an array rather than `array::IntoIter`? Because `array::IntoIter` also implements `IntoIterator<IntoIter=array::IntoIter>`

Reply:
This is enforced by the impls on `FlattenCompat`; `FlattenCompat<Map<I, F>, <U as IntoIterator>::IntoIter>` is only `TrustedLen` when `U` is a literal array.

Reply:
Ah, `U == F::Output == Map::Item` which is checked in the `I: TrustedLen<Item = ...>` bound. Ok.

I wish we could use negative where bounds to have a test that verifies this.